errbot/__init__.py

```diff
@@ -438,15 +440,22 @@ def a_webhook(self, payload):
             pass
     """
 
+    if not args:  # default uri_rule but with kwargs.
+        return lambda func: _tag_webhook(func,
+                                         _uri_from_func(func),
+                                         methods=methods,
+                                         form_param=form_param,
+                                         raw=raw)
+
     if isinstance(args[0], str):  # first param is uri_rule.
         return lambda func: _tag_webhook(func,
                                          args[0] if args[0] == '/'
                                          else args[0].rstrip('/'),  # trailing / is also be stripped on incoming.
                                          methods=methods,
                                          form_param=form_param,
                                          raw=raw)
-    return _tag_webhook(args[0],
-                        r'/' + args[0].__name__,
+    return _tag_webhook(args[0],  # naked decorator so the first parameter is a function.
+                        _uri_from_func(args[0]),
                         methods=methods,
                         form_param=form_param,
                         raw=raw)
```

Comment:
Lots of duplication going on here (only the first 1 or 2 arguments differ each time). Can we avoid this somehow?